Fix superflous shader setting updates

ShadowNinja opened the pull request:
This improves rendering performance.  The magnitude of the performance difference depends on the system, but is very substantial on the right system (~40% difference).

Rebased version of #2375 with some changes like replacing `SkyBackgroundColorProvider` with a pointer to a pointer and simplifying the `Cached*ShaderSetting` code.  The factory stuff seems to acutally be necessary to ensure that each shader has it's own `CachedShaderSetting` though.

This makes paralax mapping glitchy on leveled snow nodes (randomly switches intensity and sometimes gets a wavy effect), I'm not sure how to fix this though.  @RealBadAngel: care to take a look?

gregorycu committed:
Fix superflous shader setting updates

This improves rendering performance by ~40%

RealBadAngel commented:
@ShadowNinja sure

RealBadAngel commented:
also, this code is missing:
v3f minimap_yaw_vec = m_client->getMapper()->getYawVec();
services->setPixelShaderConstant("yawVec", (irr::f32 *)&minimap_yaw_vec, 3);

RealBadAngel commented:
After:
![screenshot_20160223_092047](https://cloud.githubusercontent.com/assets/2177790/13245685/201b927a-da0f-11e5-96b5-e58c52b66ebb.png)
Before:
![screenshot_20160223_092134](https://cloud.githubusercontent.com/assets/2177790/13245694/32383e68-da0f-11e5-8feb-4faac57aaaf1.png)

RealBadAngel commented:
Gain seems to be very small, drawtime 48 vs 50
40% seems to be propaganda ;)
But its worth it anyway. Also this code has automagically fixed some problems with transparency for shaders. Minimap become transparent as it should be. I couldnt locate a reason for that glitch.

RealBadAngel commented on src/shader.cpp:
```diff
@@ -220,47 +224,40 @@ class MainShaderConstantSetter : public IShaderConstantSetter
 		video::IVideoDriver *driver = services->getVideoDriver();
 		sanity_check(driver);
 
-		// set inverted world matrix
```
setting matrix missing in new code

ShadowNinja replied:
Yes, sofar removed these and I didn't re-add them.  They don't seem to actually have any effect.

ShadowNinja replied:
It seems neither of these matrixes are actually used by the shaders.

RealBadAngel commented on src/shader.cpp:
```diff
@@ -220,47 +224,40 @@ class MainShaderConstantSetter : public IShaderConstantSetter
 		else
 			services->setVertexShaderConstant(worldViewProj.pointer(), 4, 4);
 
-		// set transposed world matrix
```
setting matrix missing in new code

gregorycu commented:
FYI: celeron did some adhock testing with reducing the number of updates. Just a short little test.

40% is what I measured.
c55 got between 20% and 40%.

It's not the same test however. I have to be clear.

http://irc.minetest.ru/minetest-dev/2015-02-12#i_4152221
<celeron55> 50 with shaders, 70 without shaders, with shaders with patch varies between 60-70
<celeron55> so yes, there's a measurable difference

gregorycu commented:
And as you point out, I think the factory pattern makes the code as nice as it can be here. I've copped a little flak for it, so it's good that the code got validated somewhat. (By the way, I'm not a Java dev).

RealBadAngel commented:
i just got that:
![screenshot_20160223_161606](https://cloud.githubusercontent.com/assets/2177790/13255947/d771436e-da48-11e5-9739-853519d2edf7.png)

Fixer-007 commented:
WOW. Big boost for me! +20 fps ._.

cheapie commented:
While performance does seem to be increased, I get what RBA shows in his screenshot if I try to use this patch together with his VBO stuff.

ShadowNinja commented:
I fixed the snow paralax issue and the minimap, thanks RBA!
I fixed an issue with `mWorld` being set as a pixel shader setting instead of a vertex one.  Does that fix your rendering issues?

Fixer-007 commented:
Have you looked at black earth and other lightning issues?

ShadowNinja commented:
@Fixer-007: I haven't reproduced that issue, but I fixed `mWorld`, which may be the cause of it.

Fixer-007 commented:
Lightning problem is still there: https://i.imgur.com/vxq8bGg.jpg (at least with VBO), will try to test without VBO

HybridDog commented:
What were the time_of_days when you took the screenshots of the lighting bug?

Fixer-007 commented:
Interesting observation: when shadow bug kicks in on "Just Test" server, it keeps persisting after relogin, but, changing leaves to simple shadow bug goes away, I'm looking more into it.

kilbith commented:
@Fixer-007 Please replace your images by links, they're quite long to load and make harder to track the discussion. This isn't a forum.

Fixer-007 commented:
This important PR needs maintainer to fix few bugs before release.

paramat commented:
Not sure this can be in 0.4.14 due to bugs and limited dev time. We need to focus on #3848 more.

Fixer-007 commented:
@gregorycu Can you try to fix those lighting bugs if not too much trouble? :)

Fixer-007 commented:
@gregorycu Try testing on this world: [Just Test.zip](https://github.com/minetest/minetest_game/files/242731/Just.Test.zip), when I played on that server it was pretty easy (within few hours) to catch lighting errors here just with walking, changing time, relogining, going up to sky platform (some parts are captured). This world was partially captured by me, when server was running back then.

ShadowNinja commented:
I've managed to reproduce the lighting issue -- but it happens randomly and a reload fixes it.  It only affected solid nodes, snow nodeboxes were fine.  I don't really know how to debug it other than to revert chunks of code until it stops happening -- and there isn't really a good way to know if it's stopped since it's random.

Fixer-007 commented:
Funnily enough you can make it persistent after restart, really weird, try that Just Test world I've attached :) Hang out near and around tall house at -20 0 -90 try restarts, I can have black issues even after restart or at random on those positions. :}

everamzah commented:
I use this on an old laptop, OpenGL 3.3.0 Debian Stable:
`05:00.0 VGA compatible controller: NVIDIA Corporation ION VGA [GeForce 9400M] (rev b1)`

I have never seen any of those black looking nodes.  The FPS gain is very nice, although I had to resolve those shader file conflicts.

Fixer-007 commented:
Please don't let this PR rot away, it gives noticeable fps gain.

gregorycu commented:
I have rebased and fixed a bug I had identified. I am unable to reproduce the bug myself, so it would be awesome if others could test the attached PR. (Thank you for your hard work so far Fixer)

Fixer-007 commented:
@gregorycu Thanks a lot. Will start testing it today.

est31 commented:
I'm closing this in favour of #4530 as @gregorycu has rebased the change.
